base/services/w32time/ntpclient.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * PROJECT:     ReactOS Timedate Control Panel
+ * LICENSE:     GPL-2.0+ (https://spdx.org/licenses/GPL-2.0+)
+ * PURPOSE:     Queries the NTP server
+ * COPYRIGHT:   Copyright 2006 Ged Murphy <gedmurphy@gmail.com>
+ */
+
+#include "timedate.h"
+#include <winsock2.h>
+
+#define TIMEOUT 4000 /* 4 second timeout */
+
+typedef struct _INFO
+{
+    SOCKET Sock;
+    SOCKADDR_IN myAddr;
+    SOCKADDR_IN ntpAddr;
+    NTPPACKET SendPacket;
+    NTPPACKET RecvPacket;
+} INFO, *PINFO;
+
+
+static BOOL
+InitConnection(PINFO pInfo, LPSTR lpAddress)
+{
+    WSADATA wsaData;
+    HOSTENT *he;
+    INT Ret;
+
+    Ret = WSAStartup(MAKEWORD(2, 2), &wsaData);
+    if (Ret != 0)
+        return FALSE;
+
+    pInfo->Sock = socket(AF_INET,
+                         SOCK_DGRAM,
+                         0);
+    if (pInfo->Sock == INVALID_SOCKET)
+        return FALSE;
+
+    /* Setup server info */
+    he = gethostbyname(lpAddress);
+    if (he != NULL)
+    {
+        /* Setup server socket info */
+        ZeroMemory(&pInfo->ntpAddr, sizeof(SOCKADDR_IN));
+        pInfo->ntpAddr.sin_family = AF_INET; // he->h_addrtype;
+        pInfo->ntpAddr.sin_port = htons(NTPPORT);
+        pInfo->ntpAddr.sin_addr = *((struct in_addr *)he->h_addr);
+    }
+    else
+        return FALSE;
+
+    return TRUE;
+}
+
+
+static VOID
+DestroyConnection(VOID)
+{
+    WSACleanup();
+}
+
+
+static BOOL
+GetTransmitTime(PTIMEPACKET ptp)
+{
+    return TRUE;
+}
+
+
+/* Send some data to wake the server up */
+static BOOL
+SendData(PINFO pInfo)
+{
+    TIMEPACKET tp = { 0, 0 };
+    INT Ret;
+
+    ZeroMemory(&pInfo->SendPacket, sizeof(pInfo->SendPacket));
+    pInfo->SendPacket.LiVnMode = 27;
```

Comment:
What is this magic value? Any existing define for this in the PSDK?

Reply:
I do not know. I have added this comment to help clarify this value:
/* 0x1b = 011 011 - version 3 , mode 3 (client) */
Thanks.